lightning/src/ln/channelmanager.rs

```diff
@@ -433,17 +435,52 @@ pub struct ChannelManager<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref,
 	/// new channel.
 	per_peer_state: RwLock<HashMap<PublicKey, Mutex<PeerState>>>,
 
+	persistence_notifier: PersistenceNotifier,
+
 	pending_events: Mutex<Vec<events::Event>>,
 	/// Used when we have to take a BIG lock to make sure everything is self-consistent.
 	/// Essentially just when we're serializing ourselves out.
 	/// Taken first everywhere where we are making changes before any other locks.
+	/// When acquiring this lock in read mode, rather than acquiring it directly, call
+	/// `ReadLockNotifierGuard::new(..)` and pass the lock to it, to ensure its desired
+	/// callback gets executed when the lock is released.
 	total_consistency_lock: RwLock<()>,
 
 	keys_manager: K,
 
 	logger: L,
 }
 
+/// Whenever we release the `ChannelManager`'s `total_consistency_lock`, from read mode, it is
+/// desirable to notify any listeners on `wait_timeout`/`wait` that new updates are available for
+/// persistence. Therefore, this struct is responsible for locking the total consistency lock and,
+/// upon going out of scope, sending the aforementioned notification (since the lock being released
+/// indicates that the updates are ready for persistence).
+struct ReadLockNotifierGuard<'a> {
+	persistence_notifier: &'a PersistenceNotifier,
+	// We hold onto this result so the lock doesn't get released immediately.
+	_read_guard: RwLockReadGuard<'a, ()>,
+}
+
+impl<'a> ReadLockNotifierGuard<'a> {
+	fn new(lock: &'a RwLock<()>, persistence_notifier: &'a PersistenceNotifier) -> Self {
+		let read_guard = lock.read().unwrap();
+		// We want to make sure the lock is actually acquired by ReadLockNotifierGuard.
+		debug_assert!(lock.try_write().is_err());
```

Comment:
That's true, especially since I did accidentally drop the lock too early in an earlier iteration. @TheBlueMatt I'm moving the check back to where it was.